src/sicks3000.cc

```diff
@@ -310,3 +310,223 @@ unsigned short SickS3000::CreateCRC(uint8_t *Data, ssize_t length)
+  {SEG_u,SEG_9,"Traceability data incorrect or front screen calibration failed (u. - 8.)","Carry out a front screen calibration or replace the S3000, if necessary."},
+  {SEG_y,SEG__1,"Internal error in the S3000 (y. - _1.)","Replace the S3000."},
+  {SEG_y,SEG_2,"Internal I/O module error(y. - 2.)","Replace the I/O module."},
+  {SEG_y,SEG_3,"I/O module/sensor head device combination invalid (y. - 3.)","Check whether the correct I/O module has been used, and replace if necessary."},
+};
+
+int SickS3000::GetDeviceError(uint8_t first_char, uint8_t second_char, std::string* error, std::string* suggested_action)
+{
+  if(first_char != 0x00 || second_char != 0x00)
+  {
+    for(int i = 0; i < 63; i++)
+    {
+      if(first_char == s3000_errors[i].first_char && second_char == s3000_errors[i].second_char)
+      {
+        *error = s3000_errors[i].cause;
+        *suggested_action = s3000_errors[i].suggested_action;
+        return 0;
+      }
+    }
+    *error = "Invalid error code!";
+    *suggested_action = "Unable to suggest action.";
+    ROS_WARN("Invalid error code!");
+    return -1;
+  }
+  *error = "Currently there is no error.";
+  *suggested_action = "Currently there is no error.";
+  return -1;
+}
+
+int SickS3000::getDiagnosticInfo(bool host, uint16_t* scid, uint8_t* seven_seg_first_char, uint8_t* seven_seg_second_char)
+{
+  if (serial==NULL)
+  {
+    return -1;
+  }
+
+  uint8_t device_address_ = ((host) ? 0x07 : 0x08);
+  uint16_t crc = 0x0000;
+  char rx_data[256] = {0xFF};
+  int bytes_read = 0;
+  int bytes_written = 0;
+  int reply_index = 0;
+  bool read_error = true;
+
+  char interrupt_byte[1] = {0x41};
+  char reserve_token[20] = {0x00,0x00,0x41,0x44,0x19,0x00,0x00,0x05,0xFF,device_address_,0x19,0x00,0x00,0x05,0xFF,device_address_,device_address_,0x0F,0x00,0x00};
+  char release_token[20] = {0x00,0x00,0x41,0x44,0x19,0x00,0x00,0x05,0xFF,device_address_,0x19,0x00,0x00,0x05,0xFF,device_address_,0x00,0x00,0x00,0x00};
+  char request_scid_block[10] = {0x00,0x00,0x45,0x44,0x17,0x00,0x00,0x0F,0xFF,device_address_};
+  char request_block_11[10] = {0x00,0x00,0x45,0x44,0x0B,0x00,0x00,0x56,0xFF,device_address_};
+
+  crc = CreateCRC((uint8_t*)&reserve_token[10], 8);
+  reserve_token[18] = (crc & 0x00FF);
+  reserve_token[19] = (crc & 0xFF00) >> 8;
+
+  crc = CreateCRC((uint8_t*)&release_token[10], 8);
+  release_token[18] = (crc & 0x00FF);
+  release_token[19] = (crc & 0xFF00) >> 8;
+
+  //Interrupt continuous mode
+  serial->WritePort(interrupt_byte, &bytes_written, 1);
+  ros::Duration(0.1).sleep();
+
+  //Reserve token
+  serial->WritePort(reserve_token, &bytes_written, 20);
+  ros::Duration(0.1).sleep();
+
+  //Read SCID block
+  serial->WritePort(request_scid_block, &bytes_written, 10);
+  ros::Duration(0.1).sleep();
+
+  serial->BlockOnRead(100);
+  bytes_read = serial->ReadPort(rx_data);
```

Comment:
I'm concerned that with a lost byte or bitflip the driver will report wrong data or worse will hang. Can you confirm that this is not the case? For example, can we at least check the length and checksum of the returned data? eg,

```
if (bytes_read != 64)
{
  return -1;
}
```